Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Sharing/SpatialCoordinateSystemManager.cs

```diff
@@ -54,143 +36,249 @@ public class SpatialCoordinateSystemManager : Singleton<SpatialCoordinateSystemM
+                return;
+            }
+
+            string coordinateId = reader.ReadString();
+            participant.PersistentCoordinateId = coordinateId;
+            participant.Coordinate = new WorldAnchorSpatialCoordinate(coordinateId);
+        }
+
+        private async void OnLocalizeMessageReceived(SocketEndpoint socketEndpoint, string command, BinaryReader reader, int remainingDataSize)
+        {
+            if (currentLocalizationSession != null)
+            {
+                Debug.LogError($"Failed to start localization session because an existing localization session is in progress");
+                return;
+            }
+
+            Guid spatialLocalizerID = reader.ReadGuid();
+
+            if (!localizers.TryGetValue(spatialLocalizerID, out ISpatialLocalizer localizer))
+            {
+                Debug.LogError($"Request to begin localization with localizer {spatialLocalizerID} but no localizer with that ID was registered");
+                return;
+            }
+
+            if (!localizer.TryDeserializeSettings(reader, out ISpatialLocalizationSettings settings))
+            {
+                Debug.LogError($"Failed to deserialize settings for localizer {spatialLocalizerID}");
+                return;
+            }
+
+            if (!participants.TryGetValue(socketEndpoint, out SpatialCoordinateSystemParticipant participant))
+            {
+                Debug.LogError($"Could not find a SpatialCoordinateSystemParticipant for SocketEndpoint {socketEndpoint.Address}");
+                return;
+            }
+
+            await RunLocalizationSessionAsync(localizer, settings, participant);
+        }
+
+        private async Task RunLocalizationSessionAsync(ISpatialLocalizer localizer, ISpatialLocalizationSettings settings, SpatialCoordinateSystemParticipant participant)
+        {
+            using (currentLocalizationSession = localizer.CreateLocalizationSession(settings))
+            {
+                if (participant.Coordinate is WorldAnchorSpatialCoordinate worldAnchorSpatialCoordinate)
                 {
-                    StateSynchronizationObserver.Instance.UnregisterCommandHandler(command, this);
+                    worldAnchorSpatialCoordinate.RemoveAnchor();
                 }
-                if (StateSynchronizationBroadcaster.IsInitialized)
+
+                participant.IsLocatingSpatialCoordinate = true;
+                var coordinate = await currentLocalizationSession.LocalizeAsync(CancellationToken.None);
+                if (participant.PersistentCoordinateId != null)
                 {
-                    StateSynchronizationBroadcaster.Instance.UnregisterCommandHandler(command, this);
+                    participant.Coordinate = new WorldAnchorSpatialCoordinate(participant.PersistentCoordinateId, coordinate.CoordinateToWorldSpace(Vector3.zero), coordinate.CoordinateToWorldSpace(Quaternion.identity));
                 }
```

Comment:
do we feel coordinate persistence should be at the SpatialCoordinateSystemManager level compared to the participant/localizer/coordinate service? I can see some scenarios where things like ASA may have their own persistence story.

Also, with the way this is set up today, I don't know that it will work with android/ios. There is no WorldAnchorStore on iOS/Android if I understand correctly.

Reply:
I think anchor-persistence is very specific to HoloLens, and probably won't work as-is with mobile. I think that the decision to create/use a WorldAnchor or not should be controlled by the system that initiates localization in the first place. The DSLR compositor can request the connection to use the anchor, but the same app could also have a mobile participant join and that connection should not use an anchor. That's why I added a separate type of command to request using this anchor so that it can be made very specific to the DSLR scenario.

Reply:
This seems like good reasoning for keeping this persistence logic in the SpatialCoordinateSystemManager.

---
In reply to: [292215769](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/4810#discussion_r292215769) [](ancestors = 292215769)